# Proyecto final SQL - Luis Gutiérrez

In [2]:
# Conexion a la base de datos

import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Revisión de todas las tablas

In [3]:
# Tabla books

query_books = '''
                SELECT * FROM books
            
                '''

pd.io.sql.read_sql(query_books, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
# Tabla authors

query_authors = '''
                SELECT * FROM authors
            
                '''

pd.io.sql.read_sql(query_authors, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
# Tabla ratings

query_ratings = '''
            SELECT * FROM ratings
            
            '''

pd.io.sql.read_sql(query_ratings, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
# Tabla reviews

query_reviews = '''
                SELECT * FROM reviews

            '''

pd.io.sql.read_sql(query_reviews, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [7]:
# Tabla publishers

query_publishers = ''' 
                    SELECT * FROM publishers


                '''

pd.io.sql.read_sql(query_publishers, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

## Ejercicio 1

- Encuentra el número de libros publicados después del 1 de enero de 2000.

In [8]:
query1 = '''
        SELECT
            COUNT(book_id) AS book_count
        FROM
            books
        WHERE
            publication_date::DATE > '2000-01-01'
'''

pd.io.sql.read_sql(query1, con = engine)

book_count
0         819

    - Tenemos 819 libros publicados después del 01 de enero de 2000

## Ejercicio 2

- Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [9]:
query2 = '''
        SELECT
            b.book_id,
            b.title,
            r.review_count,
            rt.mean_rating
        FROM
            books b
        LEFT JOIN (
            SELECT book_id, COUNT(text) AS review_count
            FROM reviews
            GROUP BY book_id
        ) r ON b.book_id = r.book_id
        LEFT JOIN (
            SELECT book_id, AVG(rating) AS mean_rating
            FROM ratings
            GROUP BY book_id
        ) rt ON b.book_id = rt.book_id

'''

pd.io.sql.read_sql(query2, con = engine)

book_id                                              title  review_count  \
0        652          The Body in the Library (Miss Marple  #3)           2.0   
1        273                                          Galápagos           2.0   
2         51                           A Tree Grows in Brooklyn           5.0   
3        951  Undaunted Courage: The Pioneering First Missio...           2.0   
4        839                                        The Prophet           4.0   
..       ...                                                ...           ...   
995       64                                Alice in Wonderland           4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)           2.0   
997      148                                          Christine           3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)           2.0   
999      828                           The Plot Against America           2.0   

     mean_rating  
0       4.500000  
1       4.500000  
2       4.250000  
3       4.000000  
4       4.285714  
..           ...  
995     4.230769  
996     5.000000  
997     3.428571  
998     3.500000  
999     3.000000  

[1000 rows x 4 columns]

## Ejercicio 3

- Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis).

In [10]:
query3 = '''  
        SELECT
            b.publisher_id,
            p.publisher,
            COUNT(b.book_id) as book_count
        FROM
            books b
        LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
        WHERE 
            b.num_pages > 50
        GROUP BY
            b.publisher_id, p.publisher
        ORDER BY
            book_count DESC

'''

pd.io.sql.read_sql(query3, con=engine)

publisher_id                              publisher  book_count
0             212                          Penguin Books          42
1             309                                Vintage          31
2             116               Grand Central Publishing          25
3             217                       Penguin Classics          24
4              33                       Ballantine Books          19
..            ...                                    ...         ...
329           148          Harvard Business Review Press           1
330           302               Tyndale House Publishers           1
331            66                        Corgi Childrens           1
332           230  Pocket Books / Simon & Schuster  Inc.           1
333           199                          Orchard Books           1

[334 rows x 3 columns]

    - Penguin Books se lleva el trono en publicación de libros

## Ejercicio 4

- Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [11]:
query4 = '''  
        SELECT
            a.author,
            b.title,
            COUNT(rt.rating_id) AS rating_count,
            AVG(rt.rating) AS avg_rating
        FROM
            authors a
        INNER JOIN books b ON a.author_id = b.author_id
        LEFT JOIN ratings rt ON b.book_id = rt.book_id
        GROUP BY
            a.author, b.title
        HAVING
            COUNT(rt.rating_id) > 50
        ORDER BY
            avg_rating DESC

'''

pd.io.sql.read_sql(query4, con= engine)

author  \
0                          J.K. Rowling/Mary GrandPré   
1                                      J.R.R. Tolkien   
2                          J.K. Rowling/Mary GrandPré   
3                   Markus Zusak/Cao Xuân Việt Khương   
4                          J.K. Rowling/Mary GrandPré   
5                                   Louisa May Alcott   
6                          J.K. Rowling/Mary GrandPré   
7                                      J.R.R. Tolkien   
8                                       Arthur Golden   
9                                        Rick Riordan   
10                                    William Golding   
11                                          Dan Brown   
12                                      J.D. Salinger   
13           Paulo Coelho/Alan R. Clarke/Özdemir İnce   
14  William Shakespeare/Paul Werstine/Barbara A. M...   
15                                         Lois Lowry   
16            George Orwell/Boris Grabnar/Peter Škerl   
17                                          Dan Brown   
18                                    Stephenie Meyer   
19                                     John Steinbeck   

                                                title  rating_count  \
0   Harry Potter and the Prisoner of Azkaban (Harr...            82   
1   The Fellowship of the Ring (The Lord of the Ri...            74   
2   Harry Potter and the Chamber of Secrets (Harry...            80   
3                                      The Book Thief            53   
4   Harry Potter and the Half-Blood Prince (Harry ...            73   
5                                        Little Women            52   
6   Harry Potter and the Order of the Phoenix (Har...            75   
7                 The Hobbit  or There and Back Again            88   
8                                 Memoirs of a Geisha            56   
9   The Lightning Thief (Percy Jackson and the Oly...            62   
10                                  Lord of the Flies            71   
11             The Da Vinci Code (Robert Langdon  #2)            59   
12                             The Catcher in the Rye            86   
13                                      The Alchemist            57   
14                                   Romeo and Juliet            66   
15                          The Giver (The Giver  #1)            56   
16                                        Animal Farm            74   
17               Angels & Demons (Robert Langdon  #1)            84   
18                            Twilight (Twilight  #1)           160   
19                                    Of Mice and Men            61   

    avg_rating  
0     4.414634  
1     4.391892  
2     4.287500  
3     4.264151  
4     4.246575  
5     4.192308  
6     4.186667  
7     4.125000  
8     4.107143  
9     4.080645  
10    3.901408  
11    3.830508  
12    3.825581  
13    3.789474  
14    3.787879  
15    3.750000  
16    3.729730  
17    3.678571  
18    3.662500  
19    3.622951

    - J.K. Rowling y Mary GrandPré tienen la mejor calificación promedio con Harry Potter y el Prisionero de Azkabán.

## Ejercicio 5

- Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [12]:
query5 = '''  
        SELECT
            AVG(num_textreviews) AS avg_text_reviews
        FROM (
            SELECT
                r.username,
                COUNT(DISTINCT rt.rating_id) AS num_ratings,
                COUNT(DISTINCT r.review_id) AS num_textreviews
            FROM
                reviews r
            INNER JOIN ratings rt ON r.username = rt.username
            GROUP BY
                r.username
            HAVING
                COUNT(rt.rating) > 50
        ) AS user_summary;
'''

pd.io.sql.read_sql(query5, con= engine)

avg_text_reviews
0          17.45625

    - Tenemos un promedio de 17 reviews de texto para los usuarios que calificaron más de 50 libros.